python/sdist/amici/petab/petab_importer.py

```diff
@@ -0,0 +1,2282 @@
+            self._preprocess_pysb()
+        else:
+            raise AssertionError()
+
+    def _preprocess_sbml(self):
+        """Pre-process the SBML-based PEtab problem to make it
+        amici-compatible."""
+        from petab.v2.models.sbml_model import SbmlModel
+
+        if not isinstance(self.petab_problem.model, SbmlModel):
+            raise ValueError("The PEtab problem must contain an SBML model.")
+
+        # Convert petab experiments to events, because so far,
+        #  AMICI only supports preequilibration/presimulation/simulation, but
+        #  no arbitrary list of periods.
+        exp_event_conv = ExperimentsToSbmlConverter(self.petab_problem)
+        # This will always create a copy of the problem.
+        self.petab_problem = exp_event_conv.convert()
+        for experiment in self.petab_problem.experiments:
+            if len(experiment.periods) > 2:
+                # This should never happen due to the conversion above
+                raise NotImplementedError(
+                    "AMICI currently does not support more than two periods."
+                )
+
+        if self._debug:
+            print("PetabImpoter._preprocess_sbml: petab_problem:")
+            pprint(self.petab_problem.model_dump())
+            print(self.petab_problem.model.to_antimony())
+
+    def _preprocess_pysb(self):
+        """Pre-process the PySB-based PEtab problem to make it
+        amici-compatible."""
+        import pysb
+        from petab.v2.models.pysb_model import PySBModel
+
+        if not isinstance(self.petab_problem.model, PySBModel):
+            raise ValueError("The PEtab problem must contain a PySB model.")
+
+        _add_observation_model_pysb(self.petab_problem, self._jax)
+        # TODO: clarify in PEtab whether its allowed to set initial amounts
+        #  for a species without a pysb.Initial.
+        #  Currently, we fail in that case.
+        #  If so add a test case for changing the initial amount for a species
+        #  without a pysb.Initial
+        #  fixed_parameters = _add_initialization_variables(pysb_model,
+        #                                                 petab_problem)
+
+        pysb.bng.generate_equations(self.petab_problem.model.model)
+
+        # Convert PEtab v2 experiments/conditions to events
+        converter = ExperimentsToPySBConverter(self.petab_problem)
+        self.petab_problem, self._events = converter.convert()
+
+    def _upgrade_or_copy_if_needed(
+        self, problem: v1.Problem | v2.Problem
+    ) -> v2.Problem:
+        """Upgrade the problem to petab v2 if necessary.
+        Otherwise, create a deep copy of the problem."""
+        if isinstance(problem, v2.Problem):
+            return copy.deepcopy(problem)
+
+        # TODO: So far, PEtab can only upgrade file-based problems,
+        #  not petab.v1.Problem objects.
+        raise NotImplementedError(
+            "Only `petab.v2.Problem` is currently supported. "
```

Comment:
should this error message explicitly mention incompatibility with file based problems to avoid confusion?

Reply:
Reworded.